src/asl.ts

```diff
@@ -2605,152 +2841,136 @@ export class ASL {
+          ASLGraph.isConditionOutput(valueOutput) ||
+          (ASLGraph.isLiteralValue(valueOutput) &&
+            !Array.isArray(valueOutput.value))
+        ) {
+          throw new SynthError(
+            ErrorCodes.Unexpected_Error,
+            "Expected join to be performed on a variable or array constant"
+          );
+        }
 
-      if (
-        ASLGraph.isLiteralValue(separator) &&
-        typeof separator.value !== "string"
-      ) {
-        throw new SynthError(
-          ErrorCodes.Unexpected_Error,
-          "Expected join separator to be missing, undefined, a string, or a variable"
-        );
-      }
+        if (
+          ASLGraph.isLiteralValue(separator) &&
+          typeof separator.value !== "string"
+        ) {
+          throw new SynthError(
+            ErrorCodes.Unexpected_Error,
+            "Expected join separator to be missing, undefined, a string, or a variable"
+          );
+        }
 
-      // both are constants, evaluate them here.
-      if (
-        ASLGraph.isLiteralValue(valueOutput) &&
-        ASLGraph.isLiteralValue(separator)
-      ) {
-        return {
-          value: (<any[]>valueOutput.value).join(<string>separator.value),
-          containsJsonPath: false,
-        };
-      }
+        // both are constants, evaluate them here.
+        if (
+          ASLGraph.isLiteralValue(valueOutput) &&
+          ASLGraph.isLiteralValue(separator)
+        ) {
+          return {
+            value: (<any[]>valueOutput.value).join(<string>separator.value),
+            containsJsonPath: false,
+          };
+        }
 
-      const arrayPath = this.newHeapVariable();
-      const resultVariable = this.newHeapVariable();
+        const arrayPath = this.newHeapVariable();
+        const resultVariable = this.newHeapVariable();
 
-      return {
-        startState: "initArray",
-        states: {
-          // put the constant or variable array in a new temp json path
-          initArray: ASLGraph.passWithInput(
-            {
+        return {
+          startState: "initArray",
+          states: {
+            // put the constant or variable array in a new temp json path
+            initArray: ASLGraph.updateDeferredNextStates(
+              { Next: "hasNext" },
+              this.assignValue(expr.expr.expr, valueOutput, arrayPath)
+            ),
+            hasNext: {
+              Type: "Choice",
+              Choices: [
+                // not initialized and has next: init as first element
+                {
+                  ...ASL.and(
+                    ASL.isPresent(`${arrayPath}[0]`),
+                    ASL.not(ASL.isPresent(resultVariable))
+                  ),
+                  Next: "initValue",
+                },
+                // not initialized, but the array is empty
+                {
+                  ...ASL.and(
+                    ASL.not(ASL.isPresent(`${arrayPath}[0]`)),
+                    ASL.not(ASL.isPresent(resultVariable))
```

Comment:
This means empty?

Reply:
yeah, haha. Or rather, it means that there is not a valid item at index 0. It could also mean the `arrayPath` doesn't exist. In out case we know the array path exists because we created it and we don't allow undefined.

Reply:
I think you can detect an empty array by checking the `ToJson(arr) === "[]"`

Reply:
Hmm, good idea. Would take 2 states at least to check that, but it would work I think. (don't need it here, again can assume array)